### PR TITLE
Soft evict when hyperallocation of ext resource

### PR DESCRIPTION
**Pull request by shijieqin**
#### What this PR does / why we need it:
Soft evict when hyperallocation of ext resource

### CODE REVIEW

File: examples/eviction-extresource.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+apiVersion: ensurance.crane.io/v1alpha1
```
Comment (chenkaiyue) [inline examples/eviction-extresource.yaml:1]: Put this file in crane/examples/ensurance

Comment (chenkaiyue) [inline examples/eviction-extresource.yaml:1]: Put this file in crane/examples/ensurance

File: examples/extresource.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+apiVersion: ensurance.crane.io/v1alpha1
```
Comment (chenkaiyue) [inline examples/extresource.yaml:1]: Put this file in crane/examples/ensurance

Comment (chenkaiyue) [inline examples/extresource.yaml:1]: Put this file in crane/examples/ensurance

File: pkg/ensurance/collector/noderesource/noderesource.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	var distributeExtCpu int64 = 0
+	for _, pod := range pods {
+		for _, container := range pod.Spec.Containers {
+			if quantity, ok := container.Resources.Requests[corev1.ResourceName(fmt.Sprintf(noderesourceMamager.ExtResourcePrefix, corev1.ResourceCPU.String()))]; ok {
+				distributeExtCpu += quantity.Value()
+			}
+		}
+	}
+	klog.V(4).Infof("allExtCpu: %d, distributeExtCpu: %d", allExtCpu, distributeExtCpu)
+	return map[string][]common.TimeSeries{string(types.MetricNameExtCpuTotalDistribute): {{Samples: []common.Sample{{Value: (float64(distributeExtCpu) / float64(allExtCpu)) * 100, Timestamp: time.Now().Unix()}}}}}, nil
```
Comment (chenkaiyue) [inline pkg/ensurance/collector/noderesource/noderesource.go:57]: What is this used for?

Comment (chenkaiyue) [inline pkg/ensurance/collector/noderesource/noderesource.go:57]: What is this used for?

**Comment by chenkaiyue** [comment]
How to deal with the relationship between eviction-extresource and evict.
For other action, we decide how many pod should be actioned in executor module to reach the water line, in eviction-extresource this process will prepend to analyzer module.

**Comment by chenkaiyue** [comment]
How to deal with the relationship between eviction-extresource and evict.
For other action, we decide how many pod should be actioned in executor module to reach the water line, in eviction-extresource this process will prepend to analyzer module.

File: pkg/utils/pod.go [hunk 4]
```diff
@@ -85,6 +88,66 @@ func EvictPodWithGracePeriod(client clientset.Interface, pod *v1.Pod, gracePerio
...
+	if ref != nil {
+		if ref.Kind == "CloneSet" {
+			deleteLabel = "apps.kruise.io/specified-delete"
+		}
+	}
+
+	deletePath := map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"labels": map[string]interface{}{
+				deleteLabel: "true",
```
Comment (chenkaiyue) [inline pkg/utils/pod.go:103]: Add a label here, what's the next action? I think action should be accomlished.

File: pkg/utils/pod.go [hunk 5]
```diff
@@ -85,6 +88,66 @@ func EvictPodWithGracePeriod(client clientset.Interface, pod *v1.Pod, gracePerio
 	return client.CoreV1().Pods(pod.Namespace).EvictV1beta1(context.Background(), e)
 }
 
+func EvictPodForExtResource(client clientset.Interface, pod *v1.Pod) error {
+	ref := metav1.GetControllerOfNoCopy(pod)
+	deleteLabel := "gocrane.io/specified-delete"
+	if ref != nil {
+		if ref.Kind == "CloneSet" {
+			deleteLabel = "apps.kruise.io/specified-delete"
```
Comment (chenkaiyue) [inline pkg/utils/pod.go:96]: Maybe a customized logic? please delete it

**Comment by mfanjie** [comment]
fixed in https://github.com/gocrane/crane/pull/222

**Pull request closed without merging**